espnet/nets/pytorch_backend/e2e_tts_fastspeech.py

```diff
@@ -0,0 +1,686 @@
+            attention_heads=args.aheads,
+            linear_units=args.eunits,
+            num_blocks=args.elayers,
+            input_layer=encoder_input_layer,
+            dropout_rate=args.transformer_enc_dropout_rate,
+            positional_dropout_rate=args.transformer_enc_positional_dropout_rate,
+            attention_dropout_rate=args.transformer_enc_attn_dropout_rate,
+            pos_enc_class=pos_enc_class,
+            normalize_before=args.encoder_normalize_before,
+            concat_after=args.encoder_concat_after,
+            positionwise_layer_type=args.positionwise_layer_type,
+            positionwise_conv_kernel_size=args.positionwise_conv_kernel_size
+        )
+
+        # define duration predictor
+        self.duration_predictor = DurationPredictor(
+            idim=args.adim,
+            n_layers=args.duration_predictor_layers,
+            n_chans=args.duration_predictor_chans,
+            kernel_size=args.duration_predictor_kernel_size,
+            dropout_rate=args.duration_predictor_dropout_rate,
+        )
+
+        # define length regularizer
+        self.length_regularizer = LengthRegularizer()
+
+        # define decoder
+        self.decoder = Encoder(
+            idim=0,
+            attention_dim=args.adim,
+            attention_heads=args.aheads,
+            linear_units=args.dunits,
+            num_blocks=args.dlayers,
+            input_layer=None,
+            dropout_rate=args.transformer_dec_dropout_rate,
+            positional_dropout_rate=args.transformer_dec_positional_dropout_rate,
+            attention_dropout_rate=args.transformer_dec_attn_dropout_rate,
+            pos_enc_class=pos_enc_class,
+            normalize_before=args.decoder_normalize_before,
+            concat_after=args.decoder_concat_after,
+            positionwise_layer_type=args.positionwise_layer_type,
+            positionwise_conv_kernel_size=args.positionwise_conv_kernel_size
+        )
+
+        # define final projection
+        self.feat_out = torch.nn.Linear(args.adim, odim * args.reduction_factor)
+
+        # initialize parameters
+        self._reset_parameters(args)
+
+        # define teacher model
+        if args.teacher_model is not None:
+            self.teacher = self._load_teacher_model(args.teacher_model)
+        else:
+            self.teacher = None
+
+        # define duration calculator
+        if self.teacher is not None:
+            self.duration_calculator = DurationCalculator(self.teacher)
+        else:
+            self.duration_calculator = None
+
+        # transfer teacher encoder parameters
+        if args.init_encoder_from_teacher:
+            self._init_encoder_from_teacher(args.init_encoder_module)
+
+        # define criterions
+        self.duration_criterion = DurationPredictorLoss()
+        # TODO(kan-bayashi): support knowledge distillation loss
+        self.criterion = torch.nn.L1Loss()
+
+    def forward(self, xs, ilens, ys, labels, olens, *args, **kwargs):
```

Comment:
Almost the same, but in `forward` we use duration calculated by teacher model while in `inference` we use duration calculated by `duration_predictor`.

Reply:
Also, we use `duration_predictor.inference` in inference.

Reply:
I know the differences, but I thought we can do as same as https://github.com/kan-bayashi/espnet/pull/6 to avoid dup. Okay with the current code if you don't like the approach.